### PR TITLE
Coverage minima: add more fine-grained control

### PR DESCRIPTION
**Pull request by kitbellew**
Along with existing statement minimum, add branch minimum. Also, include
this pair of control at the package and file level.

### CODE REVIEW

**Comment by kitbellew** [comment]
@gslowikowski friendly reminder to review.

**Comment by kitbellew** [comment]
@gslowikowski are you interested in pursuing this still?

**Comment by gslowikowski** [comment]
Yes, I am. I didn't have enough time, when you created your pulls.

I've created some simple test projects to compare SCoverage with Cobertura and Jacoco for similar syntax constructs and it looks very different. I wasn't sure if SCoverage properly reports overall coverage.

Now I will try to find time and return to this problem. I will publish my test projects, and we will see if something in current code base should be changed before merging your pulls, or not.

**Comment by kitbellew** [comment]
@ckipp01 rebased

**Comment by jozic** [comment]
Hey @kitbellew
would you be interested in rebasing this?

**Comment by kitbellew** [comment]
> Hey @kitbellew would you be interested in rebasing this?

and them what will happen?

**Comment by jozic** [comment]
I can review and if all is good merge and release it.
I know this PR is super old, but I've only recently got the member/release permissions from @ckipp01.
@gslowikowski seems inactive, at least in this repo, and I'm willing to give back to the community by keeping this plugin maintained. At least for now. So I'm offering some help with moving this forward.
If you don't need/want this anymore, I can totally understand it.

**Comment by kitbellew** [comment]
@jozic ah, no problem, will do. (did it once before, and nothing happened subsequently, hence my question.)

**Comment by kitbellew** [comment]
@jozic сделано.

**Comment by jozic** [comment]
thanks @kitbellew 
checking it now
would you mind adding an integration test to `it` folder following the pattern there?

**Comment by kitbellew** [comment]
> would you mind adding an integration test to `it` folder following the pattern there?

done.

**Comment by kitbellew** [comment]
@jozic it might be easier to review one commit at a time, they are structured logically.

File: src/main/java/org/scoverage/plugin/SCoverageReportMojo.java [hunk 1]
```diff
@@ -547,6 +551,32 @@ else if ( !module.getPackaging().equals( "pom" ) )
         getLog().info( "Coverage aggregated reports completed." );
     }
 
+    private MavenProject findAncestorModuleWithShortestBasedir()
```
Comment (jozic) [inline src/main/java/org/scoverage/plugin/SCoverageReportMojo.java:554]: seems like this is not really related to minimums we can set, right?
seems like it's similar to https://github.com/scoverage/scoverage-maven-plugin/pull/104
could you please clarify why we need this here?  Can we extract that to a separate PR?

File: src/main/java/org/scoverage/plugin/SCoverageCheckMojo.java [hunk 2]
```diff
@@ -168,31 +223,27 @@ public void execute() throws MojoFailureException
...
                                       invokedBranchesCount, branchCount, invokedStatementCount, statementCount ) );
-        if ( minimumCoverage > 0.0 )
+
+        boolean ok = checkCoverage( getLog(), "Total", coverage,
+                                    minimumCoverage, minimumCoverageBranchTotal, true );
+        for ( MeasuredPackage pkgCoverage : CollectionConverters.asJava( coverage.packages() ) )
+        {
+            if ( !checkCoverage( getLog(), "Package:" + pkgCoverage.name(), pkgCoverage,
+                                 minimumCoverageStmtPerPackage, minimumCoverageBranchPerPackage, false ) )
+                ok = false;
```
Comment (jozic) [inline src/main/java/org/scoverage/plugin/SCoverageCheckMojo.java:235]: seems like this will continue even after after first `false`
should we instead use something like
 `CollectionConverters.asJava( coverage.packages() ).stream().allMatch`?
or do you want it to run for each package no matter what?
same question for files

File: src/main/java/org/scoverage/plugin/SCoverageCheckMojo.java [hunk 3]
```diff
@@ -201,9 +252,65 @@ else if ( coverage.statementCoveragePercent() < minimumCoverage )
...
-    private boolean is100( Double d )
+    private static boolean is100( Double d )
     {
         return Math.abs( 100 - d ) <= 0.00001d;
     }
 
-}
\ No newline at end of file
+    private static boolean checkCoverage( Log logger, String metric, CoverageMetrics metrics,
+                                          double minimumStmt, double minimimBranch, boolean logSuccessInfo )
```
Comment (jozic) [inline src/main/java/org/scoverage/plugin/SCoverageCheckMojo.java:261]: typo `minimumBranch`

File: src/main/java/org/scoverage/plugin/SCoverageCheckMojo.java [hunk 4]
```diff
@@ -201,9 +252,65 @@ else if ( coverage.statementCoveragePercent() < minimumCoverage )
...
+                                          double minimum, double actual, boolean logSuccessInfo )
+    {
+        if ( minimum <= 0 )
+        {
+            return true;
+        }
+
+        if ( is100( minimum ) && is100( actual ) )
+        {
+            logSuccess( logger, String.format( "Coverage is 100%: %s!", metric ), logSuccessInfo );
```
Comment (jozic) [inline src/main/java/org/scoverage/plugin/SCoverageCheckMojo.java:280]:
```suggestion
            logSuccess( logger, String.format( "Coverage is 100%%: %s!", metric ), logSuccessInfo );
```
otherwise it fails with `java.util.UnknownFormatConversionException: Conversion = ':'`

would be nice to add module3 to your awesome tests checking this branch )

File: src/main/java/org/scoverage/plugin/SCoverageCheckMojo.java [hunk 5]
```diff
@@ -168,31 +223,27 @@ public void execute() throws MojoFailureException
...
 
-        getLog().info( String.format( "Statement coverage.: %s%%", coverage.statementCoverageFormatted() ) );
-        getLog().info( String.format( "Branch coverage....: %s%%", coverage.branchCoverageFormatted() ) );
         getLog().debug( String.format( "invokedBranchesCount:%d / branchCount:%d, invokedStatementCount:%d / statementCount:%d",
                                       invokedBranchesCount, branchCount, invokedStatementCount, statementCount ) );
-        if ( minimumCoverage > 0.0 )
+
+        boolean ok = checkCoverage( getLog(), "Total", coverage,
+                                    minimumCoverage, minimumCoverageBranchTotal, true );
+        for ( MeasuredPackage pkgCoverage : CollectionConverters.asJava( coverage.packages() ) )
```
Comment (jozic) [inline src/main/java/org/scoverage/plugin/SCoverageCheckMojo.java:231]: this could be a large list, right? for files it will be even larger
 how about wrapping this into `if(minimumCoverageStmtPerFile > 0.0 && minimumCoverageBranchPerPackage > 0.0)` to avoid checking every package/file with default/0 minimum?

File: src/it/test_coverage_minima/validate.groovy [hunk 6]
```diff
@@ -0,0 +1,38 @@
+def checkModule(logText, module, coverageLog) {
```
Comment (jozic) [inline src/it/test_coverage_minima/validate.groovy:1]: Awesome tests! Thank you.

**Review by jozic — COMMENTED** [review]
looks very nice, left some questions/suggestions

File: src/it/test_coverage_minima/validate.groovy (hunk 6 above)
Comment (kitbellew) [reply src/it/test_coverage_minima/validate.groovy:1]: you added the tests, i just duplicated :)

File: src/main/java/org/scoverage/plugin/SCoverageCheckMojo.java [hunk 7]
```diff
@@ -81,6 +86,56 @@ public class SCoverageCheckMojo
     @Parameter( property = "scoverage.minimumCoverage", defaultValue = "0" )
     private Double minimumCoverage;
 
+    /**
+     * Required minimum total branch coverage.
+     * <br>
+     * <br>
+     * See <a href="https://github.com/scoverage/sbt-scoverage#minimum-coverage">https://github.com/scoverage/sbt-scoverage#minimum-coverage</a> for additional documentation.
+     * <br>
```
[the remark this answers is not in the document]
Comment (kitbellew) [reply src/main/java/org/scoverage/plugin/SCoverageCheckMojo.java:94]: will do

File: src/main/java/org/scoverage/plugin/SCoverageCheckMojo.java (hunk 5 above)
Comment (kitbellew) [reply src/main/java/org/scoverage/plugin/SCoverageCheckMojo.java:231]: will do

File: src/main/java/org/scoverage/plugin/SCoverageCheckMojo.java (hunk 2 above)
Comment (kitbellew) [reply src/main/java/org/scoverage/plugin/SCoverageCheckMojo.java:235]: good question. we can either report until first failure (in which case the user would fix something and then find out there's more), or report everything. options:
- report everything
- report only first failure
- put the choice under yet another parameter.

File: src/main/java/org/scoverage/plugin/SCoverageCheckMojo.java (hunk 3 above)
Comment (kitbellew) [reply src/main/java/org/scoverage/plugin/SCoverageCheckMojo.java:261]: will do

File: src/main/java/org/scoverage/plugin/SCoverageCheckMojo.java (hunk 4 above)
Comment (kitbellew) [reply src/main/java/org/scoverage/plugin/SCoverageCheckMojo.java:280]: will do

File: src/main/java/org/scoverage/plugin/SCoverageReportMojo.java (hunk 1 above)
Comment (kitbellew) [reply src/main/java/org/scoverage/plugin/SCoverageReportMojo.java:554]: ah, correct. the code is what we used in our production for many years, and it's likely that there were two unrelated changes to make everything work. will remove from this pr (and the linked one might be a better implementation).

File: src/main/java/org/scoverage/plugin/SCoverageCheckMojo.java (hunk 2 above)
Comment (jozic) [reply src/main/java/org/scoverage/plugin/SCoverageCheckMojo.java:235]: I personally generate a report on a failure, to see what's not covered. So the first failure seems fine to me. 
But I can also see a point of having all the packages/files checked at once.

How about we do fail-fast in this PR and then we can do a separate PR to add an extra flag, like `failFast` with `true` by default?

File: src/main/java/org/scoverage/plugin/SCoverageReportMojo.java (hunk 1 above)
Comment (jozic) [reply src/main/java/org/scoverage/plugin/SCoverageReportMojo.java:554]: Just a heads-up, i've merged #104, so dropping this particular change should be safe now.

File: src/main/java/org/scoverage/plugin/SCoverageCheckMojo.java (hunk 2 above)
Comment (kitbellew) [reply src/main/java/org/scoverage/plugin/SCoverageCheckMojo.java:235]: @jozic good morning and happy new year.

i finally got around to working on this, and here's what i realized: we will have reporting on all packages and files if it's successful but will stop at the first error if it happens (with a single [ERROR] message after possibly a bunch of [INFO]/[DEBUG]). which is a bit weird since the opposite is probably of more interest.

so, i see several alternatives:
- report everything, just like now
- report everything, except use [DEBUG] for second and subsequent errors
- collect all info/debug messages until the first error; if there are no errors, only then log the info/debug.

thoughts?

Comment (jozic) [reply src/main/java/org/scoverage/plugin/SCoverageCheckMojo.java:235]: Thanks @kitbellew, happy New Year to you as well!

Good point, i guess reporting packages and files as debug for success and error for failure is a fair game. So we should probably leave it as you have it )

Comment (kitbellew) [reply src/main/java/org/scoverage/plugin/SCoverageCheckMojo.java]: @jozic all done and pushed.

**Comment by kitbellew** [comment]
perhaps windows tests will now succeed. added `.replaceAll("\r\n", "\n")`.

File: src/it/test_coverage_minima/validate.groovy [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    assert new File(basedir, module + "/target/scoverage.xml").exists()
+    assert new File(basedir, module + "/target/site/scoverage/index.html").exists()
+    def entry = logText.find {
+        it.startsWith("scoverage-maven-plugin:") &&
```
Comment (jozic) [inline src/it/test_coverage_minima/validate.groovy:5]: i've noticed that with maven 3.9.6 ( i have that one locally) this check fails, as seems like maven 3.9 changed format from `[INFO] --- scoverage-maven-plugin:2.0.1-SNAPSHOT:check (default-cli) @ module01 ---`
to `[INFO] --- scoverage:2.0.1-SNAPSHOT:check (default-cli) @ module01 ---`
so changing to `it.startsWith("scoverage") &&` seems fine for both 3.9.x and 3.8.x

**Review by jozic — COMMENTED** [review]
Almost there, let's make the test maven 3.9 compatible and we should be good to merge and then hopefully release soon after

File: src/it/test_coverage_minima/validate.groovy (hunk 8 above)
Comment (kitbellew) [reply src/it/test_coverage_minima/validate.groovy:5]: will do

File: src/main/java/org/scoverage/plugin/SCoverageCheckMojo.java [hunk 9]
```diff
@@ -168,31 +233,19 @@ public void execute() throws MojoFailureException
...
 
-        getLog().info( String.format( "Statement coverage.: %s%%", coverage.statementCoverageFormatted() ) );
-        getLog().info( String.format( "Branch coverage....: %s%%", coverage.branchCoverageFormatted() ) );
         getLog().debug( String.format( "invokedBranchesCount:%d / branchCount:%d, invokedStatementCount:%d / statementCount:%d",
                                       invokedBranchesCount, branchCount, invokedStatementCount, statementCount ) );
-        if ( minimumCoverage > 0.0 )
+
+        boolean ok = checkCoverage( getLog(), "Total", coverage,
+                                    minimumCoverage, minimumCoverageBranchTotal, true );
+        ok = checkCoverage( getLog(), "Package:", coverage.packages(), x -> x.name(),
```
[the remark this answers is not in the document]
Comment (kitbellew) [reply src/main/java/org/scoverage/plugin/SCoverageCheckMojo.java:241]: will do. not really familiar with new java syntax, more of a scala guy myself :)

File: src/main/java/org/scoverage/plugin/SCoverageCheckMojo.java [hunk 10]
```diff
@@ -201,9 +254,89 @@ else if ( coverage.statementCoveragePercent() < minimumCoverage )
...
+    private static <T> boolean checkAll( scala.collection.Iterable<T> iterable, Predicate<T> predicate )
+    {
+        boolean ok = true;
+        for ( T elem : CollectionConverters.asJava( iterable ) )
+        {
+            if ( !predicate.test( elem ) )
+            {
+                ok = false;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (kitbellew) [reply src/main/java/org/scoverage/plugin/SCoverageCheckMojo.java:338]: will do, thank you, aligns well with what i did above.

**Comment by kitbellew** [comment]
@jozic all done, ptal. happy new year :)

File: src/main/java/org/scoverage/plugin/SCoverageCheckMojo.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (jozic) [reply src/main/java/org/scoverage/plugin/SCoverageCheckMojo.java]: same here, I just don't like `x` variables and Intellij was kind/smart enough to suggest those :)

**Review by jozic — APPROVED** [review]
lgtm

@kitbellew thanks for you contribution and for your patience during reviews and those almost 6 years :)

I see that @gslowikowski wanted to do some extra checks, but as he's no longer active we gonna move forward and merge/release this. We can always revisit things later if needed.

Happy New Year!

**Comment by jozic** [comment]
@kitbellew this is released in 2.0.1
